static/app/views/releases/thresholdsList/thresholdGroupRows.tsx

```diff
@@ -0,0 +1,413 @@
+    return byId;
+  }, [thresholds]);
+
+  const thresholdIdSet = useMemo(() => {
+    return new Set([
+      ...thresholds.map(threshold => threshold.id),
+      ...Object.keys(editingThresholds),
+    ]);
+  }, [thresholds, editingThresholds]);
+
+  const initializeNewThreshold = () => {
+    const thresholdId = `${NEW_THRESHOLD_PREFIX}-${newThresholdIterator}`;
+    const [windowValue, windowSuffix] = parseLargestSuffix(defaultWindow);
+    const newThreshold: EditingThreshold = {
+      id: thresholdId,
+      project,
+      environment,
+      windowValue,
+      windowSuffix,
+      threshold_type: 'total_error_count',
+      trigger_type: 'over',
+      value: 0,
+      hasError: false,
+    };
+    const updatedEditingThresholds = {...editingThresholds};
+    updatedEditingThresholds[thresholdId] = newThreshold;
+    setEditingThresholds(updatedEditingThresholds);
+    setNewThresholdIterator(newThresholdIterator + 1);
+  };
+
+  const enableEditThreshold = thresholdId => {
+    const updatedEditingThresholds = {...editingThresholds};
+    const threshold = JSON.parse(JSON.stringify(thresholdsById[thresholdId]));
+    const [windowValue, windowSuffix] = parseLargestSuffix(threshold.window_in_seconds);
+    updatedEditingThresholds[thresholdId] = {
+      ...threshold,
+      windowValue,
+      windowSuffix,
+      hasError: false,
+    };
+    setEditingThresholds(updatedEditingThresholds);
+  };
+
+  const saveThreshold = thresholdIds => {
+    thresholdIds.forEach(id => {
+      const thresholdData = editingThresholds[id];
+      const submitData: {[key: string]: any} = {...thresholdData};
+      let path = `/projects/${orgSlug}/${thresholdData.project.slug}/release-thresholds/${id}/`;
+      let method: APIRequestMethod = 'PUT';
+      if (id.includes(NEW_THRESHOLD_PREFIX)) {
+        path = `/projects/${orgSlug}/${thresholdData.project.slug}/release-thresholds/`;
+        method = 'POST';
+      }
+      const seconds = moment
+        // @ts-ignore: ts is unhappy with moment string duration format
+        .duration(thresholdData.windowValue, thresholdData.windowSuffix)
+        .as('seconds');
+      submitData.window_in_seconds = seconds;
+      submitData.environment = thresholdData.environment.name; // api expects environment as a string
+      const request = api.requestPromise(path, {
+        method,
+        data: submitData,
+      });
+      request
+        .then(() => {
+          refetch();
+          closeEditForm(id);
+        })
+        .catch(_err => {
+          setError('Issue saving threshold');
+          const errorThreshold = {
+            ...submitData,
+            hasError: true,
+            environment: thresholdData.environment, // convert local state environment back to object
+          };
+          const updatedEditingThresholds = {...editingThresholds};
+          updatedEditingThresholds[id] = errorThreshold as EditingThreshold;
+          setEditingThresholds(updatedEditingThresholds);
```

Comment:
this might need to use the current state since editingThresholds could be out of date when the promise finally resolves
```suggestion
          setEditingThresholds(editingThresholds => {
                    const errorThreshold = {
				            ...submitData,
				            hasError: true,
				            environment: thresholdData.environment, // convert local state environment back to object
				          };
				          const updatedEditingThresholds = {...editingThresholds};
				          updatedEditingThresholds[id] = errorThreshold as EditingThreshold;
         });
```

Reply:
i'm not sure why github is using tabs today 🤷

Reply:
Ah nice.
👍